source/lac/trilinos_precondition.cc

```
@@ -955,8 +959,258 @@ namespace TrilinosWrappers
+        parameter_list.set("null space: dimension",
+                           distributed_constant_modes.NumVectors());
+        if (my_size > 0)
+          parameter_list.set("null space: vectors",
+                             distributed_constant_modes.Values());
+        // We need to set a valid pointer to data even if there is no data on
+        // the current processor. Therefore, pass a dummy in that case
+        else
+          parameter_list.set("null space: vectors",
+                             &dummy[0]);
+      }
+
+    initialize (matrix, parameter_list);
+  }
+
+
+
+  void
+  PreconditionAMGMueLu::initialize (const SparseMatrix           &matrix,
+                                    Teuchos::ParameterList &muelu_parameters)
+  {
+    initialize(matrix.trilinos_matrix(), muelu_parameters);
+  }
+
+
+
+  void
+  PreconditionAMGMueLu::initialize (const Epetra_CrsMatrix       &matrix,
+                                    Teuchos::ParameterList &muelu_parameters)
+  {
+    // We cannot use MueLu::CreateEpetraOperator directly because, we cannot
+    // transfer ownership of MueLu::EpetraOperator from Teuchos::RCP to
+    // std::shared_ptr.
+
+    // For now, just use serial node, i.e. no multithreaing or GPU.
+    typedef KokkosClassic::DefaultNode::DefaultNodeType node;
+    preconditioner.reset ();
+
+    // Cast matrix into a MueLu::Matrix. The constness needs to be cast away.
+    // MueLu uses Teuchos::RCP which are Trilinos version of std::shared_ptr.
+    Teuchos::RCP<Epetra_CrsMatrix> rcp_matrix = Teuchos::rcpFromRef(
+                                                  *(const_cast<Epetra_CrsMatrix *>(&matrix)));
+    Teuchos::RCP<Xpetra::CrsMatrix<double,int,int,node> > muelu_crs_matrix =
+      Teuchos::rcp(new Xpetra::EpetraCrsMatrix (rcp_matrix));
+    Teuchos::RCP<Xpetra::Matrix<double,int,int,node> > muelu_matrix =
+      Teuchos::rcp(new Xpetra::CrsMatrixWrap<double,int,int,node> (muelu_crs_matrix));
+
```

What happens if you compile deal.II with 64bit integers, @Rombur ? Does this preconditioner work if the matrix is small enough? If not, do we need to throw an exception?